### DISCUSSION TITLE
Experimental | urlbuilder

### DISCUSSION THREAD

**Discussion by garrettladley** [a-h/templ#867] [Ideas]
Hello everyone, there's a new experimental urlbuilder feature in [templ-go/x](https://github.com/templ-go/x).

Checkout the PR [here](https://github.com/templ-go/x/pull/2).

We're looking to collect community feedback on the feature! Does it meet your needs? If not, what else would be necessary?

Thanks again, G

**Comment by mac-hel** [comment]
Is there a way to build relative url?

**Reply by joerdav** [reply]
Currently no, only URLs that contain a scheme and a host work at the moment. But good suggestion, I think that would be useful!

**Reply by acaloiaro** [reply]
I came to suggest the same. 

This PR (https://github.com/templ-go/x/pull/5) is the API that I prefer.

**Reply by acaloiaro** [reply]
@joerdav I'm curious if you have any feedback on the API design of https://github.com/templ-go/x/pull/5

**Reply by joerdav** [reply]
I understand the design choices, but I think the fact we are circumventing `"net/url"` for most of the construction of the URL we are missing some of the assurances that the std library gives us.

I think that potentially the safest way to design this builder is to wrap a `*url.URL`:

```go
type URLBuilder struct {
	url *url.URL
	err      error // allows error handling to be deferred.
}
```

Then we can provide a `urlbuilder.New` func, as you already have done. But the builder functions modify the url directly:

```go
func (u *URLBuilder) Fragment(fragment string) *URLBuilder {
	u.url.Fragment = fragment
	return u
}

// urlbuilder.New().Path("user", userid)
func (u *URLBuilder) Path(base string, elements ...string) *URLBuilder {
	path, err := url.JoinPath(base, elements...)
	u.err = err
	u.url.Path = path
	return u
}
``` 

It could also be beneficial to provide a "Parse" function that allows passing of an initial URL. It could set the `err` value that would be returned in the final `Build` step.

**Reply by acaloiaro** [reply]
I agree, `*url.URL` should be a core implementation detail. 

As an API, my goal of commenting here is to remove the current scheme/host requirement for those who have a need to build relative URLs. 

Just to clarify @garrettladley came up with this nice little API. I'm simply advocating for its design, if not its current implementation details.

**Comment by Nigel2392** [comment]
Maybe I'm missing something, and I hate to be _that guy_, but what is the use case for this?

Would a custom format function that returns a valid url object not suffice?

To me, _at first glance_, it looks shiny and cool, but not very practical.
